Source/Core/Core/HW/WiimoteEmu/MotionPlus.cpp

```diff
@@ -0,0 +1,376 @@
+// Copyright 2019 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "Core/HW/WiimoteEmu/MotionPlus.h"
+
+#include "Common/BitUtils.h"
+#include "Common/ChunkFile.h"
+#include "Common/Logging/Log.h"
+#include "Common/MsgHandler.h"
+
+namespace WiimoteEmu
+{
+MotionPlus::MotionPlus() : Extension("MotionPlus")
+{
+}
+
+void MotionPlus::Reset()
+{
+  reg_data = {};
+
+  constexpr std::array<u8, 6> initial_id = {0x00, 0x00, 0xA6, 0x20, 0x00, 0x05};
+
+  // FYI: This ID changes on activation
+  std::copy(std::begin(initial_id), std::end(initial_id), reg_data.ext_identifier);
+
+  // TODO: determine meaning of calibration data:
+  static const u8 cdata[32] = {
+      0x78, 0xd9, 0x78, 0x38, 0x77, 0x9d, 0x2f, 0x0c, 0xcf, 0xf0, 0x31,
+      0xad, 0xc8, 0x0b, 0x5e, 0x39, 0x6f, 0x81, 0x7b, 0x89, 0x78, 0x51,
+      0x33, 0x60, 0xc9, 0xf5, 0x37, 0xc1, 0x2d, 0xe9, 0x15, 0x8d,
+  };
+
+  std::copy(std::begin(cdata), std::end(cdata), reg_data.calibration_data);
+
+  // TODO: determine the meaning behind this:
+  static const u8 cert[64] = {
+      0x99, 0x1a, 0x07, 0x1b, 0x97, 0xf1, 0x11, 0x78, 0x0c, 0x42, 0x2b, 0x68, 0xdf,
+      0x44, 0x38, 0x0d, 0x2b, 0x7e, 0xd6, 0x84, 0x84, 0x58, 0x65, 0xc9, 0xf2, 0x95,
+      0xd9, 0xaf, 0xb6, 0xc4, 0x87, 0xd5, 0x18, 0xdb, 0x67, 0x3a, 0xc0, 0x71, 0xec,
+      0x3e, 0xf4, 0xe6, 0x7e, 0x35, 0xa3, 0x29, 0xf8, 0x1f, 0xc5, 0x7c, 0x3d, 0xb9,
+      0x56, 0x22, 0x95, 0x98, 0x8f, 0xfb, 0x66, 0x3e, 0x9a, 0xdd, 0xeb, 0x7e,
+  };
+
+  std::copy(std::begin(cert), std::end(cert), reg_data.cert_data);
+}
+
+void MotionPlus::DoState(PointerWrap& p)
+{
+  p.Do(reg_data);
+}
+
+bool MotionPlus::IsActive() const
+{
+  return ACTIVE_DEVICE_ADDR << 1 == reg_data.ext_identifier[2];
+}
+
+MotionPlus::PassthroughMode MotionPlus::GetPassthroughMode() const
+{
+  return static_cast<PassthroughMode>(reg_data.ext_identifier[4]);
+}
+
+ExtensionPort& MotionPlus::GetExtPort()
+{
+  return m_extension_port;
+}
+
+int MotionPlus::BusRead(u8 slave_addr, u8 addr, int count, u8* data_out)
+{
+  if (IsActive())
+  {
+    // Motion plus does not respond to 0x53 when activated
+    if (ACTIVE_DEVICE_ADDR == slave_addr)
+      return RawRead(&reg_data, addr, count, data_out);
+    else
+      return 0;
+  }
+  else
+  {
+    if (INACTIVE_DEVICE_ADDR == slave_addr)
+      return RawRead(&reg_data, addr, count, data_out);
```

Comment:
This should either be braced since the else case is braced (or just don't even join them together, since both branches consist of returns).

Reply:
👍